lib/liquid/tags/for.rb

```diff
@@ -128,7 +128,14 @@ def collection_segment(context)
       end
 
       collection = context.evaluate(@collection_name)
-      collection = collection.to_a if collection.is_a?(Range)
+      if collection.is_a?(Range)
+        score_limit = context.resource_limits.render_score_limit
+        if score_limit && score_limit < collection.size
+          context.resource_limits.render_score += collection.size
+          raise MemoryError.new("Memory limits exceeded".freeze)
```

Comment:
The general idea here is correct, but this will cause a subtle behavior difference. The semantics are that rendering should continue _until_ the render score is reached - meaning the first `0..RENDER_SCORE` items should be rendered before the memory error.

Instead, we should do something like:

```ruby
if score_limit && score_limit < collection.size
  collection = collection.begin..score_limit
end
```

If the render score is too high still, we may need to adjust `Liquid::Utils.slice_collection` so that it doesn't accidentally `to_a` to range.